src/IO/S3/Client.cpp

```diff
@@ -267,17 +267,32 @@ void Client::setKMSHeaders(RequestType & request) const
     }
 }
 
+
+template <typename RequestType>
```

Comment:
Rewrite without templates.

`SetAdditionalCustomHeaderValue` is a virtual method.